README
General Information
===================

This is AceConn version 1.1.0. AceConn is a C library for accessing
the socket protocol version of the AceDB servers.

The official AceDB ftp site is:
  ftp://ftp.sanger.ac.uk/pub/acedb

The official web site is:
  http://www.acedb.org/

Installation
============

See the file 'INSTALL'

Notes about AceConn 1.1.0
=========================

* This most recent version of AceConn now builds/installs using the
  gnu autoconf system.

How to report bugs
==================

Bugs should be reported to the acedb bug tracking system. 
Send an email to acedb-bug@sanger.ac.uk.

In the bug report please include:

* Information about your system. For instance:

   - What operating system and version
   - For Linux, what version of the C library

  And anything else you think is relevant.

* How to reproduce the bug. 

  If you can reproduce it with one of the test programs that are built 
  in the tests/ subdirectory, that will be most convenient.  Otherwise, 
  please include a short test program that exhibits the behavior. 
  As a last resort, you can also provide a pointer to a larger piece 
  of software that can be downloaded.

* If the bug was a crash, the exact text that was printed out
  when the crash occured.

* Further information such as stack traces may be useful, but
  is not necessary.

Patches
=======

Patches should be submitted to acedb@sanger.ac.uk. If the
patch fixes an existing bug, add the patch as an attachment
to that bug report.

Otherwise, enter a new bug report that describes the patch,
and attach the patch to that bug report.

Bug reports containing patches should include the PATCH keyword
in their keyword fields. If the patch adds to or changes the GLib 
programming interface, the API keyword should also be included.

Patches should be in unified diff form. (The -u option to GNU
diff.)



NEWS
Overview of Changes from GLib 2.12.8 to GLib 2.12.9
===================================================

* AceConn moved to autoconf-based build system.


test/simpletest.c
/*  File: simpletest.c
 *  Author: Ed Griffiths (edgrif@sanger.ac.uk)
 *  Copyright (c) Sanger Institute, 2002
 *-------------------------------------------------------------------
 * Acedb is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * or see the on-line version at http://www.gnu.org/copyleft/gpl.txt
 *-------------------------------------------------------------------
 * Description: Simple test program for socket package allowing a single
 *              command to be sent to the server and then quitting.
 * HISTORY:
 * Last edited: Sep  5 14:46 2003 (edgrif)
 * Created: Mon Mar 11 14:25:09 2002 (edgrif)
 * CVS info:   $Id: simpletest.c,v 1.1 2007-10-03 13:28:00 edgrif Exp $
 *-------------------------------------------------------------------
 */

#include <stdlib.h>
#include <stdio.h>

#include <getopt.h>
#include <glib.h>
#include <AceConn.h>

/* Defaults, ok for me, not for anyone else....                              */
#define DEF_HOST     "griffin2"
#define DEF_PORT     22222
#define DEF_USERID   "anonymous"
#define DEF_PASSWD   "anyone"
#define DEF_TIMEOUT  120

static void usage(char *err_str) ;



int main(int argc, char **argv)
{
  int rc = EXIT_SUCCESS ;
  guint version, release, update ;
  AceConnStatus status ;
  char *host = DEF_HOST ;
  int port = DEF_PORT ;
  char *userid = DEF_USERID ;
  char *passwd = DEF_PASSWD ;
  int timeout = DEF_TIMEOUT ;
  AceConnection connection = NULL ;
  char *request = "find sequence" ;
  char *reply = NULL ;
  int reply_len ;

  while (1)
    {
      struct option long_options[] =
      {
	{"host",     required_argument, NULL, 'h'},
	{"port",     required_argument, NULL, 'p'},
	{"userid",   required_argument, NULL, 'u'},
	{"passwd",   required_argument, NULL, 'x'},
	{"timeout",  required_argument, NULL, 't'},
	{0, 0, 0, 0}
      } ;
      int c ;
      int option_index = 0 ;				    /* getopt_long stores the option index */
							    /* here. */

      c = getopt_long_only(argc, argv, "t:p:u:x:oif:",
			   long_options, &option_index);

      /* Detect the end of the options. */
      if (c == -1)
        break;

      switch (c)
        {
        case 0:
          /* If this option set a flag, do nothing else now. */
          if (long_options[option_index].flag != 0)
            break;
          printf ("option %s", long_options[option_index].name);
          if (optarg)
            printf (" with arg %s", optarg);
          printf ("\n");
          break;

	case 'h':
	  host = optarg ;
	  break ;

        case 'p':
	  if (sscanf(optarg, "%d", &port) != 1)
	    {
	      usage(g_strdup_printf("Bad -p option: %s\n", optarg)) ;
	    }
          break;

	case 'u':
	  userid = optarg ;
	  break ;

	case 'x':
	  passwd = optarg ;
	  break ;

        case 't':
	  if (sscanf(optarg, "%d", &timeout) != 1)
	    {
	      usage(g_strdup_printf("Bad -t option: %s\n", optarg)) ;
	    }
          break;

        case '?':
          /* getopt_long prints an error message. */
	  usage(NULL) ;
          break;

        default:
	  usage(g_strdup_printf("Internal logic error, unexpected value returned by"
				"getopt_long_only() call: %c\n", c)) ;
        }

    }


  /* User can enter a single string request to be sent to server.            */
  if (optind < argc)
    {
      request = argv[optind++] ;
      if (optind < argc)
	usage(g_strdup_printf("Excess command line parameter: %s\n", argv[optind])) ;
    }

  /* Show the version.                                                       */
  printf("\nAceConn Header  Version: %d.%d.%d\n",
	 ACECONN_VERSION, ACECONN_RELEASE, ACECONN_UPDATE) ;
  AceConnGetVersion(&version, &release, &update) ;
  printf("AceConn Library Version: %d.%d.%d\n", version, release, update) ;
  if (!ACECONN_CHECK_VERSION(version, release, update))
    usage("AceConn.h and libAceConn.a are incompatible.\n") ;
  else
    printf("\n") ;


  /* Make the connection.                                                    */
  if ((status = AceConnCreate(&connection, host, port, userid, passwd, timeout)) != ACECONN_OK)
    {
      printf("Cannot create initial connection object.\n") ;

      exit(EXIT_FAILURE) ;
    }

  /* Connect to server.                                                      */
  if (status == ACECONN_OK)
    {
      status = AceConnConnect(connection) ;
      if (status != ACECONN_OK)
	{
	  printf("Error connecting: %s\n", AceConnGetErrorMsg(connection, status)) ;
	}
    }

  if (status == ACECONN_OK)
    {
      status = AceConnRequest(connection, request, (void *)(&reply), &reply_len) ; 
      if (status == ACECONN_OK || status == ACECONN_QUIT)
	{
	  printf("Reply length: %d\n", reply_len) ;
	  printf("Reply text: >|%s|<\n", reply) ;
	  g_free(reply) ;
	}
      else
	{
	  printf("Error making request: %s\n", AceConnGetErrorMsg(connection, status)) ;
	}
    }

  if (status == ACECONN_OK || status == ACECONN_QUIT)
    {
      status = AceConnDisconnect(connection) ;
      if (status != ACECONN_OK)
	{
	  printf("Error disconnnecting: %s\n", AceConnGetErrorMsg(connection, status)) ;
	}
    }


  return rc ;
}



static void usage(char *err_str)
{
  if (err_str)
    printf("Error: %s\n", err_str) ;

  printf("\nusage:\n"
	 "\tsimpletest [-h host] [-p number] [-u userid] [-x passwd][-t time_out_secs] {request]\n"
	 "Allows you to send a single request to the Acedb socket server and then quits.\n") ;

  exit(EXIT_FAILURE) ;

  return ;
}


test/simpleclient.c
/*  File: simpleclient.c
 *  Author: Ed Griffiths (edgrif@sanger.ac.uk)
 *  Copyright (c) Sanger Institute, 2002
 *-------------------------------------------------------------------
 * Acedb is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * or see the on-line version at http://www.gnu.org/copyleft/gpl.txt
 *-------------------------------------------------------------------
 * Description: Test program for socket package allowing commands to be
 *              sent to server and replies to be displayed until
 *              user types "quit".
 * HISTORY:
 * Last edited: Dec 13 12:11 2004 (edgrif)
 * Created: Mon Mar 11 14:25:09 2002 (edgrif)
 * CVS info:   $Id: simpleclient.c,v 1.1 2007-10-03 13:28:00 edgrif Exp $
 *-------------------------------------------------------------------
 */

#include <stdlib.h>
#include <stdio.h>

#include <getopt.h>
#include <readline/readline.h>
#include <glib.h>
#include <AceConnProto.h>
#include <AceConn.h>

/* Defaults, ok for me, not for anyone else....                              */
#define DEF_HOST     "griffin2"
#define DEF_PORT     22222
#define DEF_USERID   "anonymous"
#define DEF_PASSWD   "anyone"
#define DEF_TIMEOUT  120
#define DEF_PROMPT   "Ace-Conn > "


static char *prog_name_G = NULL ;


static void usage(char *err_str) ;



int main(int argc, char **argv)
{
  int rc = EXIT_SUCCESS ;
  AceConnStatus status ;
  char *host = DEF_HOST ;
  int port = DEF_PORT ;
  char *userid = DEF_USERID ;
  char *passwd = DEF_PASSWD ;
  int timeout = DEF_TIMEOUT ;
  gboolean stats_only = FALSE ;
  AceConnection connection = NULL ;
  char *reply = NULL ;
  int reply_len ;
  gboolean finished ;
  char *prompt = DEF_PROMPT ;

  prog_name_G = argv[0] ;

  while (1)
    {
      struct option long_options[] =
      {
	{"host",      required_argument, NULL, 'h'},
	{"port",      required_argument, NULL, 'p'},
	{"userid",    required_argument, NULL, 'u'},
	{"passwd",    required_argument, NULL, 'x'},
	{"timeout",   required_argument, NULL, 't'},
	{"statsonly", no_argument,       NULL, 's'},
	{0, 0, 0, 0}
      } ;
      int c ;
      int option_index = 0 ;				    /* getopt_long stores the option index */
							    /* here. */

      c = getopt_long_only(argc, argv, "h:p:u:x:t:s",
			   long_options, &option_index);

      /* Detect the end of the options. */
      if (c == -1)
        break;

      switch (c)
        {
        case 0:
          /* If this option set a flag, do nothing else now. */
          if (long_options[option_index].flag != 0)
            break;
          printf ("option %s", long_options[option_index].name);
          if (optarg)
            printf (" with arg %s", optarg);
          printf ("\n");
          break;

	case 'h':
	  host = optarg ;
	  break ;

        case 'p':
	  if (sscanf(optarg, "%d", &port) != 1)
	    {
	      usage(g_strdup_printf("Bad -p option: %s\n", optarg)) ;
	    }
          break;

	case 'u':
	  userid = optarg ;
	  break ;

	case 'x':
	  passwd = optarg ;
	  break ;

        case 't':
	  if (sscanf(optarg, "%d", &timeout) != 1)
	    {
	      usage(g_strdup_printf("Bad -t option: %s\n", optarg)) ;
	    }
          break;

	case 's':
	  stats_only = TRUE ;
	  break ;

        case '?':
          /* getopt_long prints an error message. */
	  usage(NULL) ;
          break;

        default:
	  usage(g_strdup_printf("Internal logic error, unexpected value returned by"
				"getopt_long_only() call: %c\n", c)) ;
        }

    }


  /* Shouldn't be anything else on command line.                             */
  if (optind < argc)
    {
      usage(g_strdup_printf("Excess command line parameter: %s\n", argv[optind])) ;
    }

  /* Make the connection.                                                    */
  if ((status = AceConnCreate(&connection, host, port, userid, passwd, timeout)) != ACECONN_OK)
    {
      printf("Cannot create initial connection object.\n") ;

      exit(EXIT_FAILURE) ;
    }

  /* Connect to server.                                                      */
  if (status == ACECONN_OK)
    {
      status = AceConnConnect(connection) ;
      if (status != ACECONN_OK)
	{
	  char *err_msg = AceConnGetErrorMsg(connection, status) ;
	  
	  printf("Error connecting: %s\n", err_msg) ;
	}
    }

  /* loop reading user input and relaying stuff to server and then printing  */
  /* replies.                                                                */
  if (status == ACECONN_OK)
    {
      finished = FALSE ;
      while (!finished)
	{
	  char *request = NULL ;

	  request = readline(prompt) ;
	  if (!request)
	    printf("\n") ;
	  else
	    {
	      status = AceConnRequest(connection, request, (void *)(&reply), &reply_len) ; 
	      if (status == ACECONN_OK || status == ACECONN_QUIT)
		{
		  if (stats_only)
		    printf("Reply length: %d\n", reply_len) ;
		  else
		    {
		      if (reply_len > 0)
			{
			  printf("%s\n", reply) ;
			  g_free(reply) ;
			}
		    }

		  if (status == ACECONN_QUIT)
		    finished = TRUE ;
		}
	      else
		{
		  char *errmsg = AceConnGetErrorMsg(connection, status) ;

		  printf("Error making request: %s\n", errmsg) ;

		  finished = TRUE ;
		}
	      free(request) ;
	    }
	}
    }


  /* Close connection to server.                                             */
  if (status == ACECONN_OK)
    {
      status = AceConnDisconnect(connection) ;
      if (status != ACECONN_OK)
	{
	  printf("Error disconnnecting: %s\n", AceConnGetErrorMsg(connection, status)) ;
	}
    }


  return rc ;
}



static void usage(char *err_str)
{
  if (err_str)
    printf("\nError: %s\n", err_str) ;

  printf("\n"
	 "usage:\n"
	 "\t%s [-h host][-p number][-u userid][-x passwd][-t time_out_secs][-s]  [request]\n"
	 "Allows you to send a single request to the Acedb socket server and then quits.\n\n"
	 "\t-t\ta timeout of zero means \"never timeout\"\n"
	 "\t-s\twill show message lengths only\n"
	 "\n",
	 prog_name_G) ;

  exit(EXIT_FAILURE) ;

  return ;
}


src/aceconn.c
/*  File: aceconn.c
 *  Author: Ed Griffiths (edgrif@sanger.ac.uk)
 *  Copyright (c) Sanger Institute, 2002
 *-------------------------------------------------------------------
 * Acedb is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * or see the on-line version at http://www.gnu.org/copyleft/gpl.txt
 *-------------------------------------------------------------------
 * Description: A C package for accessing the acedb socket server.
 *              
 *              The intention is that this package will be standalone
 *              and not dependent on the rest of acedb so anyone can
 *              take it and use it.
 *              
 *              The code makes use of glib which is an GNU licensed
 *              package of utility functions.
 *              
 *              This file contains just the interface functions for the
 *              package.
 *              
 * Exported functions: See AceConn.h
 * HISTORY:
 * Last edited: Dec 13 11:49 2004 (edgrif)
 * Created: Wed Mar  6 13:55:37 2002 (edgrif)
 * CVS info:   $Id: aceconn.c,v 1.1 2007-10-03 13:28:00 edgrif Exp $
 *-------------------------------------------------------------------
 */

#include <glib.h>
#include <aceconn_.h>


static AceConnStatus doServerClose(AceConnection connection) ;



/**@name Interface Functions
 * The AceConn package has a simple set of interface functions that allow
 * you to open a connection to an acedb server, send requests and receive
 * answers. */

/*@{ Start of interface functions. */


/** Create the an unconnected connection control block.
 *
 * @return ACECONN_OK if connect to server successful, otherwise
 * status indicates what failure was.
 *
 * @param connection_out returned valid connection (opaque type),
 *        untouched unless create succeeds.
 * @param server_netid host network id of server
 * @param server_port servers port on host
 * @param userid user known to acedb server
 * @param passwd users password (will be encrypted before sending).
 * @param timeout time to wait for reply from server before returning
 *        control to application
 */
AceConnStatus AceConnCreate(AceConnection *connection_out,
			    char *server_netid, int server_port,
			    char *userid, char *passwd, int timeout)
{
  AceConnStatus status = ACECONN_OK ;
  AceConnection connection = NULL ;

  /* Could check port number for  server_port < 1024 || server_port > 65535,
   * but probably over kill. */
  if (!connection_out || server_netid == NULL || server_port <= 0
      || userid == NULL || passwd == NULL)
    status = ACECONN_BADARGS ;
  else
    {
      if (!(connection = aceconnCreate(server_netid, server_port, userid, passwd, timeout)))
	status = ACECONN_NOCREATE ;
      else
	*connection_out = connection ;
    }

  return status ;
}

/** Open a connection to the server. Does all the hand shaking necessary to
 * say "hello" to the server, after this acedb requests can be sent to the
 * server.
 *
 * @return ACECONN_OK if connect to server successful, otherwise
 * status indicates what failure was.
 *
 * @param connection_inout server connection
 */
AceConnStatus AceConnConnect(AceConnection connection_inout)
{
  AceConnStatus status = ACECONN_OK ;

  if (connection_inout == NULL) 
    {
      status = ACECONN_INVALIDCONN ;
    }

  if (status == ACECONN_OK)
    {
      status = aceconnServerOpen(connection_inout) ;
    }

  if (status == ACECONN_OK)
    {
      status = aceconnServerHandshake(connection_inout) ;
    }

  /* If there was a problem, then clear up. */
  if (status != ACECONN_OK)
    {
      doServerClose(connection_inout) ;
    }

  return status ;
}



/** Sends an acedb request to the server and then waits for and returns
 * the reply from the server, note that the reply can be NULL if the user
 * has set "quiet" mode on and the request is a command with no response.
 *
 * @return ACECONN_OK if request sent and reply retrieved successfully, otherwise
 * status indicates what failure was.
 *
 * @param connection_out server connection
 * @param request, a C string containing the acedb request, e.g. "find sequence *"
 * @param reply_out, the reply received from the server, <B>note</B> this may be either
 * a C string or binary data as when a postscript image is returned or NULL.
 * @param reply_len_out the length of the data returned, <B>note</B> that for a C string,
 * this <b>includes</b> the terminating null char of the string, if reply_out is NULL
 * then this will be zero.
 */
AceConnStatus AceConnRequest(AceConnection connection,
			     char *request,
			     void **reply_out, int *reply_len_out)
{
  AceConnStatus status = ACECONN_OK ;

  if (!aceconnIsValid(connection))
    {
      status = ACECONN_INVALIDCONN ;
    }

  if (status == ACECONN_OK)
    {
      if (request == NULL || reply_out == NULL || reply_len_out == NULL)
	{
	  status = ACECONN_BADARGS ;
	  aceconnSetErrMsg(connection, ACECONN_BADARGS, "%s", "Bad args to function AceConnRequest()") ;
	}
    }

  if (status == ACECONN_OK)
    {
      status = aceconnServerRequest(connection, request, reply_out, reply_len_out) ;

      /* For a normal request we just send it, but if the request was "quit" */
      /* then we close the connection to the server as well, note that if    */
      /* the close was successful we return a status of "quit" so the caller */
      /* can detect this situation.                                          */
      if (status == ACECONN_OK && (strcmp(request, ACECONN_CLIENT_GOODBYE) == 0))
	{
	  connection->state = ACECONNSTATE_OPEN ;
	  status = aceconnServerClose(connection) ;
	  if (status == ACECONN_OK)
	    {
	      status = ACECONN_QUIT ;
	      aceconnSetErrMsg(connection, ACECONN_QUIT, "%s",
			       "Connection to server closed by \"quit\" request.") ;
	    }
	}
    }

  return status ;
}


/** Closes the connection to an acedb server. If close is
 * successful then connection is freed, otherwise it is not
 * to allow application to get the error message.
 *
 * @return ACECONN_OK if disconnected successfully, otherwise
 * status indicates what failure was.
 *
 * @param connection server connection (opaque type)
 */
AceConnStatus AceConnDisconnect(AceConnection connection)
{
  AceConnStatus status = ACECONN_OK ;

  if (!aceconnIsValid(connection))
    status = ACECONN_INVALIDCONN ;

  if (status == ACECONN_OK)
    {
      status = doServerClose(connection) ;
    }

  return status ;
}


/** Simple free of a connection struct.
 *
 * @return nothing
 *
 * @param connection server connection (opaque type)
 */
void AceConnDestroy(AceConnection connection)
{
  aceconnDestroy(connection) ;

  return ;
}

/*@} End of interface functions. */


/**@name Utility Functions
 * Following are all basically utility functions that supplement the Interface
 * functions. */

/*@{ Start of utility functions. */


/** Returns the current version of the AceConn library.
 *
 * @return nothing.
 *
 * @param version, release, update, the version, release and update
 * are returned in these parameters.
 */
void AceConnGetVersion(guint *version, guint *release, guint *update)
{
  if (version)
    *version = ACECONN_VERSION ;
  if (release)
    *release = ACECONN_RELEASE ;
  if (update)
    *update = ACECONN_UPDATE ;

  return ;
}


/** Sets the timeout for all communication with the server. If the server has
 * not responded after "timeout" seconds AceConn returns control to the
 * application, setting a status of ACECONN_TIMEDOUT.
 *
 * @return ACECONN_OK if timeout set correctly, or ACECONN_INVALIDCONN
 * if connection is invalid.
 *
 * @param connection_out server connection (opaque type)
 * @param timeout the new timeout in seconds.
 */
AceConnStatus AceConnSetTimeout(AceConnection connection, int timeout)
{
  AceConnStatus status = ACECONN_OK ;

  if (!aceconnIsValid(connection))
    status = ACECONN_INVALIDCONN ;

  if (status == ACECONN_OK)
    {
      connection->timeout = timeout ;
      status = ACECONN_OK ;
    }

  return status ;
}


/** Checks whether connection to server is open.
 *
 * @return ACECONN_OK if connection is open or ready, ACECONN_INVALIDCONN
 * if connection is invalid otherwise ACECONN_NOTOPEN.
 *
 * @param connection_out server connection (opaque type)
 */
AceConnStatus AceConnConnectionOpen(AceConnection connection)
{
  AceConnStatus status = ACECONN_NOTOPEN ;

  if (!aceconnIsValid(connection))
    status = ACECONN_INVALIDCONN ;
  else if (connection->state != ACECONNSTATE_NONE)
    status = ACECONN_OK ;

  return status ;
}


/** If the connection is valid and contains a last error status
 * that matches err_status then a string is returned containing
 * a more detailed description of the last error detected by the
 * AceConn library. Otherwise a simple text string version of
 * err_status is returned.
 * This call can be used for all errors returned by the AceConn
 * library, e.g. the call works if connection is NULL, invalid
 * or valid.
 * Can be used to obtain a simple text string version of any
 * error status by calling with  connection == NULL.
 * The returned string is internal to the AceConn package and
 * the caller should not alter or free it.
 *
 * @return error string, or NULL if err_status == ACECONN_OK
 *
 * @param connection server connection (opaque type)
 * @param err_status last error status returned by AceConn package
 */
char *AceConnGetErrorMsg(AceConnection connection, AceConnStatus err_status)
{
  char *err_msg = NULL ;

  if (aceconnIsValid(connection) && connection->last_err_status == err_status)
    err_msg = connection->last_errmsg ;
  else
    err_msg = aceconnGetSimpleErrMsg(err_status) ;

  return err_msg ;
}

/** DEPRECATED FUNCTION, use AceConnGetErrorMsg() instead.
 *
 * Returns a string containing a description of the last error
 * encountered by the library, i.e. call will only return a string if
 * an interface call has returned something other than ACECONN_OK and
 * connection is still valid. The returned string is internal to the
 * AceConn package and the caller should not alter or free it.
 *
 * @return error string, or NULL if there has been no error or
 * the connection is invalid.
 *
 * @param connection_out server connection (opaque type)
 */
char *AceConnGetLastErrMsg(AceConnection connection)
{
  char *err_msg = NULL ;

  if (aceconnIsValid(connection))
    err_msg = connection->last_errmsg ;

  return err_msg ;
}

/*@} End of utility functions. */



/*                                                                           */
/* Internal functions.                                                       */
/*                                                                           */

static AceConnStatus doServerClose(AceConnection connection)
{
  AceConnStatus status = ACECONN_OK ;

  if (status == ACECONN_OK && connection->state != ACECONNSTATE_NONE)
    {
      status = aceconnServerClose(connection) ;
    }

  return status ;
}



src/aceconn_.h
/*  File: aceconn_.h
 *  Author: Ed Griffiths (edgrif@sanger.ac.uk)
 *  Copyright (c) Sanger Institute, 2002
 *-------------------------------------------------------------------
 * Acedb is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * or see the on-line version at http://www.gnu.org/copyleft/gpl.txt
 *-------------------------------------------------------------------
 * Description: Private header for C package for accessing the AceDB
 *              socket server.
 *              
 *              Functions that have external binding but are internal
 *              to this package all start "aceconn" to separate them
 *              from the interface functions that start "AceConn".
 * HISTORY:
 * Last edited: Sep  5 11:29 2003 (edgrif)
 * Created: Wed Mar  6 13:58:41 2002 (edgrif)
 * CVS info:   $Id: aceconn_.h,v 1.1 2007-10-03 13:28:00 edgrif Exp $
 *-------------------------------------------------------------------
 */
#ifndef DEF_ACECONN_P_H
#define DEF_ACECONN_P_H

#include <stdio.h>					    /* Needed for NULL surprisingly... */
#include <string.h>
#include <glib.h>					    /* Needed here for some basic types. */
#include <AceConnProto.h>
#include <AceConn.h>


/* Opaque handle to an message passed to server.                             */
typedef struct AceConnMsgRec *AceConnMsg ;


/* Connection can be in one of a number of states.                           */
typedef enum
{
  ACECONNSTATE_NONE,					    /* not connected at all to server. */
  ACECONNSTATE_OPEN,					    /* socket open. */
  ACECONNSTATE_READY					    /* connection ready for requests. */
} AceConnState ;


/* Holds all the data for a particular connection to the ace server.         */
typedef struct _AceConnRec
{
  char **valid ;					    /* Used for checking valid block. */

  AceConnState state ;

  /* Server details                                                          */
  char *host ;
  int  port ;
  int  socket ;
  int  timeout ;					    /* command timeout in secs. */

  /* User details                                                            */
  char *userid ;
  char *passwd ;
  char *passwd_hash ;
  char *nonce_hash ;

  /* Current request. */
  AceConnMsg msg ;

  /* Misc. */
  AceConnStatus last_err_status ;			    /* status describiing last error. */
  gchar *last_errmsg ;					    /* string describing last error. */


} AceConnRec ;


/* Basic creation/destruction of an AceConnRec.                              */
AceConnection aceconnCreate(char *server_netid, int server_port,
			    char *userid, char *passwd, int timeout) ;
void aceconnDestroy(AceConnection connection) ;


/* Server requests.                                                          */
AceConnStatus aceconnServerOpen(AceConnection connection) ;
AceConnStatus aceconnServerHandshake(AceConnection connection) ;
AceConnStatus aceconnServerRequest(AceConnection connection,
				   char *request, void **reply, int *reply_len) ;
AceConnStatus aceconnServerClose(AceConnection connection) ;


/* Utilities.                                                                */

gboolean aceconnIsValid(AceConnection connection) ;

/* Used to get simple string version of error status. */
char *aceconnGetSimpleErrMsg(AceConnStatus status) ;

/* Use to set an error message in the connection struct, application can     */
/* retrieve error message for display/logging etc.                           */
/* N.B. for simple strings do this:  aceconnSetErrMsg(conn, "%s", string) ;  */
void aceconnSetErrMsg(AceConnection connection, AceConnStatus status,
		      char *format, ...) ;
void aceconnFreeErrMsg(AceConnection connection) ;



#endif /* !defined DEF_ACECONN_P_H */


include/AceConn.h
/*  File: AceConn.h
 *  Author: Ed Griffiths (edgrif@sanger.ac.uk)
 *  Copyright (c) Sanger Institute, 2002
 *-------------------------------------------------------------------
 * Acedb is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * or see the on-line version at http://www.gnu.org/copyleft/gpl.txt
 *-------------------------------------------------------------------
 * Description: Interface for communicating with the AceDB server.
 *              Currently the server is sockets based but this is not
 *              exposed in this interface.
 *              
 *              You should see AceConn.html for full details
 *              of the interface.
 *              
 * HISTORY:
 * Last edited: Nov 12 15:31 2003 (edgrif)
 * Created: Wed Mar  6 13:58:41 2002 (edgrif)
 * CVS info:   $Id: AceConn.h,v 1.1 2007-10-03 13:28:00 edgrif Exp $
 *-------------------------------------------------------------------
 */
#ifndef DEF_ACECONN_H
#define DEF_ACECONN_H

#include <glib.h>

/* Library version stuff.                                                    */
#define     ACECONN_VERSION 1
#define     ACECONN_RELEASE 1
#define     ACECONN_UPDATE  0

#define ACECONN_CHECK_VERSION(version, release, update)    \
    (ACECONN_VERSION > (version) || \
     (ACECONN_VERSION == (version) && ACECONN_RELEASE > (release)) || \
     (ACECONN_VERSION == (version) && ACECONN_RELEASE == (release) && \
      ACECONN_UPDATE >= (update)))


/* Opaque handle to a connection to an Acedb socket server.                  */
typedef struct _AceConnRec *AceConnection ;

/* Return codes for the AceConn calls.                                       */
typedef enum {ACECONN_OK = 0,
	      ACECONN_QUIT,				    /* Command to server was "quit" so
							       connection is now closed. */
	      ACECONN_INVALIDCONN,			    /* Connection points to invalid memory. */
	      ACECONN_BADARGS,				    /* caller has supplied bad args. */
	      ACECONN_NOTOPEN,				    /* Connection not open. */
	      ACECONN_NOALLOC,				    /* could not allocate  */
	      ACECONN_NOSOCK,				    /* socket creation problem. */
	      ACECONN_UNKNOWNHOST,			    /* server host not known. */
	      ACECONN_NOCONNECT,			    /* Could not connect to host. */
	      ACECONN_NOSELECT,				    /* select on socket failed. */
	      ACECONN_HANDSHAKE,			    /* Handshake to server failed. */
	      ACECONN_READERROR,			    /* Error in reading from socket. */
	      ACECONN_WRITEERROR,			    /* Error in writing to socket. */
	      ACECONN_SIGSET,				    /* Problem with signal setting. */
	      ACECONN_NONBLOCK,				    /* Non-blocking for socket failed. */
	      ACECONN_TIMEDOUT,				    /* Connection timed out. */
	      ACECONN_NOCREATE,				    /* Could not create connection
							       control block. */
	      ACECONN_INTERNAL				    /* Dire, internal package error. */
} AceConnStatus ;


/* Default time (seconds) to wait for reply from server. */
enum {ACECONN_DEFAULT_TIMEOUT = 120} ;


/* Creates a connection struct. */
AceConnStatus AceConnCreate(AceConnection *connection, char *server_netid, int server_port,
			    char *userid, char *passwd, int timeout) ;


/* Open a connection to the database. */
AceConnStatus AceConnConnect(AceConnection connection) ;

/* Send a request to the database and get the reply. */
AceConnStatus AceConnRequest(AceConnection connection,
			     char *request,
			     void **reply, int *reply_len) ;

/* Close down the connection. */
AceConnStatus AceConnDisconnect(AceConnection connection) ;


/* Free the connection struct. */
void AceConnDestroy(AceConnection connection) ;



/* Utility functions.                                                        */
/*                                                                           */

void AceConnGetVersion(guint *version, guint *release, guint *update) ;

AceConnStatus AceConnSetTimeout(AceConnection connection, int timeout) ;

AceConnStatus AceConnConnectionOpen(AceConnection connection) ;


char *AceConnGetErrorMsg(AceConnection connection, AceConnStatus err_status) ;


/* Deprecated, use AceConnGetErrorMsg() instead. */
char *AceConnGetLastErrMsg(AceConnection connection) ;


#endif /* !defined DEF_ACECONN_H */
